### PR TITLE
added submodule support

### PR DESCRIPTION
**Pull request by japettyjohn**
Added option to init and update submodules. If depth is specified it uses shallow submodules.

### CODE REVIEW

**Comment by thockin** [comment]
I don't see anything wrong with this, but I don't know submodules well.  Can you add a set of cases to the test_e2e.sh to demonstrate this?

**Comment by thockin** [comment]
also, you need to sign the CLA :)

**Comment by stp-ip** [comment]
From a submodule perspective it looks good. Waiting for some test cases and the CLA.

**Comment by abh** [comment]
FWIW this is working well for me -- I made a build at `quay.io/abh/git-sync:v2.0.5-submodule`

**Comment by japettyjohn** [comment]
@abh That's great to hear. I've been running with success as well for some months, I haven't gotten around to implementing shallow checkouts though which would help a lot for large/ancient dependencies and make it consistent with the top level checkout features.

**Comment by thockin** [comment]
Sorry for letting this languish.  It looks overall ok, but it needs an e2e test, please?

File: cmd/git-sync/main.go [hunk 1]
```diff
@@ -54,6 +54,8 @@ var flWait = flag.Float64("wait", envFloat("GIT_SYNC_WAIT", 0),
 	"the number of seconds between syncs")
 var flOneTime = flag.Bool("one-time", envBool("GIT_SYNC_ONE_TIME", false),
 	"exit after the initial checkout")
+var flSubmodule = flag.Bool("submodule", envBool("GIT_SYNC_SUBMODULE", false),
```
Comment (thockin) [inline cmd/git-sync/main.go]: should the flag be "submodules" ?

**Comment by abh** [comment]
@japettyjohn any chance of resurrecting this so it can be merged? :-)

**Comment by StephenWithPH** [comment]
@abh ... I'd like to hear more about how it's working for you.

Background: I took a stab at adding E2E tests to cover @japettyjohn 's work. The code works as expected in the repo itself, but it does *not* work down the symlinked-path. My tests are failing, since I'm expecting the submodule in the symlinked path.

As I understand it, the symlinked path is what is supposed to matter; from the docs:
>When it re-pulls, it updates the destination directory atomically. In order to do this, it uses a git worktree in a subdirectory of the --root and flips a symlink

In the tree below, `ottherrepo` is submoduled inside of `repo`. The code runs and does pull it down (`main.go:465] run("/tmp/git-sync-test.14210/root/rev-e9fc17a57997bad70aafe925b73054100ef61878"): git submodule update --init)`, but none of this makes it to the `worktree` where it counts (`main.go:465] run("/tmp/git-sync-test.14210/root"): git worktree add /tmp/git-sync-test.14210/root/rev-e9fc17a57997bad70aafe925b73054100ef61878 origin/master`).

```
git-sync-test.14210 $ tree
.
├── log.branch-sync
├── log.cross-branch-tag-sync
├── log.default-sync
├── log.head-once
├── log.head-sync
├── log.rev-once
├── log.rev-sync
├── log.submodule
├── log.tag-sync
├── otherrepo
│   └── otherfile
├── repo
│   ├── file
│   └── otherrepo
│       └── otherfile
└── root
    ├── repo -> rev-e9fc17a57997bad70aafe925b73054100ef61878
    └── rev-e9fc17a57997bad70aafe925b73054100ef61878
        └── file
```

https://git-scm.com/docs/git-worktree tells me that:
>BUGS
Multiple checkout in general is still experimental, and the support for submodules is incomplete. It is NOT recommended to make multiple checkouts of a superproject.

**Pull request closed without merging**